pkg/distributor/distributor.go

```diff
@@ -166,55 +171,57 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
 		}
 	}
 
-	replicationFactor.Set(float64(ring.ReplicationFactor()))
+	replicationFactor.Set(float64(ingestersRing.ReplicationFactor()))
 	cfg.PoolConfig.RemoteTimeout = cfg.RemoteTimeout
 
 	replicas, err := newClusterTracker(cfg.HATrackerConfig)
 	if err != nil {
 		return nil, err
 	}
 
-	d := &Distributor{
-		cfg:            cfg,
-		ring:           ring,
-		ingesterPool:   ingester_client.NewPool(cfg.PoolConfig, ring, cfg.ingesterClientFactory, util.Logger),
-		billingClient:  billingClient,
-		limits:         limits,
-		ingestLimiters: map[string]*rate.Limiter{},
-		quit:           make(chan struct{}),
-		Replicas:       replicas,
-	}
-	go d.loop()
-
-	return d, nil
-}
+	// Create the configured ingestion rate limit strategy (local or global). In case
+	// it's an internal dependency and can't join the distributors ring, we skip rate
+	// limiting.
+	var ingestionRateStrategy limiter.RateLimiterStrategy
+	var distributorsRing *ring.Lifecycler
 
-func (d *Distributor) loop() {
-	if d.cfg.LimiterReloadPeriod == 0 {
-		return
-	}
+	if !canJoinRing {
```

Comment:
Can be simplified to:

```
	if !canJoinRing {
		ingestionRateStrategy = newLocalIngestionRateStrategy(limits)
	} else if limits.IngestionRateStrategy() == validation.GlobalIngestionRateStrategy {
		distributorsRing, err = ring.NewLifecycler(cfg.DistributorRing.ToLifecyclerConfig(), nil, "distributor", ring.DistributorRingKey)
		if err != nil {
			return nil, err
		}

		distributorsRing.Start()

		ingestionRateStrategy = newGlobalIngestionRateStrategy(limits, distributorsRing)
	}
```

Reply:
I think this means that it's never going to be infinite, maybe double the logic once?

Reply:
Not sure. I think the current logic is safer and the intention is more clear (when it's an internal dependency, do not configure any rate limit at all).